### DISCUSSION TITLE
Functions to calculate preconditioners for linear solvers

### DISCUSSION THREAD

**Discussion by MetLab-HKUST** [jax-ml/jax#23055] [General]
Hi, it is great that JAX has already implemented a few linear solvers, such as jax.scipy.sparse.linalg.bicgstab. However, those solvers can converge faster with a preconditioner, but JAX does not have them. Does anyone know any implementation in JAX, such as 'spilu'? Could the JAX team provide an official version?

The following page shows what is available in FEniCS: https://fenicsproject.org/pub/tutorial/html/._ftut1017.html#ftut:app:solver:prec

Thanks.

**Comment by f0uriest** [comment]
Most "standard" preconditioners like incomplete LU/cholesky rely on having access to the elements of the operator, usually as a sparse array, which are currently still experimental in JAX, so I don't know how feasible it would be to implement something like that.

**Reply by MetLab-HKUST** [reply]
Indeed. I realized that sometimes, we only have a linear operator function instead of the specific sparse matrix. This makes calculating the precondition challenging.
